plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerFacetImpl.java

```diff
@@ -0,0 +1,589 @@
+    // FIXME Nexus is changing content type from application/json to text/plain even if is a json file
+    tx.setBlob(metadataAsset, metadataAssetName, streamSupplier.get(), hashAlgorithms, null, payload.getContentType());
+    tx.saveAsset(metadataAsset);
+  }
+
+  @Override
+  @Transactional(retryOn = IllegalStateException.class)
+  public Payload getLayer(final String imageId) throws IOException {
+    validateImageId(imageId);
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    String metadataAssetName = imageId + "/json";
+    String layerAssetName = imageId + "/layer";
+
+    Bucket bucket = tx.getBucket();
+    Asset metadataAsset = findAsset(tx, bucket, metadataAssetName);
+    if (metadataAsset == null || Boolean.TRUE.equals(metadataAsset.formatAttributes().get(P_IN_PROGRESS))) {
+      throw new V1ApiException("Image not found", HttpStatus.NOT_FOUND);
+    }
+    Asset layerAsset = findAsset(tx, bucket, layerAssetName);
+    if (layerAsset == null) {
+      throw new V1ApiException("Image not found", HttpStatus.NOT_FOUND);
+    }
+
+    return marshall(layerAsset, tx.requireBlob(layerAsset.requireBlobRef()));
+  }
+
+  @Override
+  public void putLayer(final String imageId, final Payload content) throws IOException, InvalidContentException {
+    validateImageId(imageId);
+    try (final TempStreamSupplier streamSupplier = new TempStreamSupplier(content.openInputStream())) {
+      doPutLayer(imageId, streamSupplier, content);
+    }
+  }
+
+  @Transactional(retryOn = {ONeedRetryException.class, ORecordDuplicatedException.class})
+  protected void doPutLayer(final String imageId, final Supplier<InputStream> streamSupplier, final Payload payload)
+      throws IOException
+  {
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    String metadataAssetName = imageId + "/json";
+    String layerAssetName = imageId + "/layer";
+
+    Bucket bucket = tx.getBucket();
+    Asset metadataAsset = findAsset(tx, bucket, metadataAssetName);
+    if (metadataAsset == null) {
+      throw new V1ApiException("Image not found", HttpStatus.NOT_FOUND);
+    }
+    if (!Boolean.TRUE.equals(metadataAsset.formatAttributes().get(P_IN_PROGRESS))) {
+      throw new V1ApiException("Image already exists", HttpStatus.CONFLICT);
+    }
+    Asset layerAsset = findAsset(tx, bucket, layerAssetName);
+    if (layerAsset == null) {
+      layerAsset = tx.createAsset(bucket, getRepository().getFormat());
+      layerAsset.name(layerAssetName);
+      layerAsset.formatAttributes().set(P_PATH, "/v1/images/" + layerAssetName);
+    }
+
+    AttributesMap contentAttributes = null;
+    if (payload instanceof Content) {
+      contentAttributes = ((Content) payload).getAttributes();
+    }
+    Content.applyToAsset(layerAsset, Content.maintainLastModified(layerAsset, contentAttributes));
+    // FIXME Nexus is changing content type from application/octet-stream to application/gzip
```

Comment:
This is a "good thing". Octet stream is actually "no idea", while Tika does realise that layers are TGZ